src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalMemberEditAnalysis.cs

```diff
@@ -0,0 +1,368 @@
+            {
+                // We are performing semantic span-based analysis for member-only edit scenario.
+                // Instead of computing the analyzer diagnostics for the entire document,
+                // we have computed the new diagnostics just for the edited member span.
+                Debug.Assert(executor.AnalysisScope.Span.HasValue);
+                Debug.Assert(executor.AnalysisScope.Span.Value == changedMember.FullSpan);
+                Debug.Assert(diagnostics.All(d => !d.HasTextSpan || changedMember.FullSpan.IntersectsWith(d.GetTextSpan())));
+
+                // We now try to get the new document diagnostics by performing an incremental update:
+                //   1. Re-using all the old cached diagnostics outside the edited member node from a prior
+                //      document snapshot, but with updated diagnostic spans.
+                //      AND
+                //   2. Replacing old diagnostics for the edited member node in a prior document snapshot
+                //      with the new diagnostics for this member node in the latest document snaphot.
+                // If we are unable to perform this incremental diagnostics update,
+                // we fallback to computing the diagnostics for the entire document.
+                var ranges = savedMemberRangesForSpanBasedAnalyzers[stateSet.Analyzer]!;
+                if (TryGetUpdatedDocumentDiagnostics(existingData, ranges.Value.Ranges, diagnostics.AsImmutableOrEmpty(), changedMember.SyntaxTree, changedMember, changedMemberId, out var updatedDiagnostics))
+                {
+#if DEBUG_INCREMENTAL_ANALYSIS
+                    await ValidateMemberDiagnosticsAsync(executor, stateSet, updatedDiagnostics, cancellationToken).ConfigureAwait(false);
+#endif
+                    return updatedDiagnostics;
+                }
+                else
+                {
+                    // Incremental diagnostics update failed.
+                    // Fallback to computing the diagnostics for the entire document.
+                    var documentExecutor = executor.With(executor.AnalysisScope.WithSpan(null));
+                    return await computeAnalyzerDiagnosticsAsync(stateSet.Analyzer, documentExecutor, cancellationToken).ConfigureAwait(false);
+                }
+
+#if DEBUG_INCREMENTAL_ANALYSIS
+                static async Task ValidateMemberDiagnosticsAsync(DocumentAnalysisExecutor executor, StateSet stateSet, ImmutableArray<DiagnosticData> diagnostics, CancellationToken cancellationToken)
+                {
+                    executor = executor.With(executor.AnalysisScope.WithSpan(null));
+                    var expected = await executor.ComputeDiagnosticsAsync(stateSet.Analyzer, cancellationToken).ConfigureAwait(false);
+                    Debug.Assert(diagnostics.SetEquals(expected));
+                }
+#endif
+            }
+
+            private static bool TryGetUpdatedDocumentDiagnostics(
```

Comment:
This is the core logic that computes latest diagnostics using latest changed member diagnostics and cached diagnostics from document snapshot prior to member edit. Almost all the code in the method is borrowed from prior implementation. I primarily added additional location handling here as we were either dropping them or not updating the spans for those, both of them were leading to asserts.